lib/Importer/ProtobufLoader.cpp

```diff
@@ -399,4 +402,67 @@ Expected<TypeRef> ProtobufLoader::loadQuantTy(const std::string &name,
   return mod_.uniqueType(k, dims, dummyScale, thisUniqueOffset);
 }
 
+void ProtobufLoader::updateOperatorInputsPrecision(
+      std::unordered_map<std::string, NodeValue> &inputsNodeValueMap,
+      std::unordered_map<std::string, NodeValue> &nonConstantsNodeValueMap) {
+  for (auto itr = inputsNodeValueMap.begin();
+            itr != inputsNodeValueMap.end(); ++itr){
+    std::string inputName = itr->first;
+    NodeValue input = itr->second;
+    // If input type is fp32 add convert to fp16 node and update
+    // Global nodeValueByName_ map, so that operator in the model
+    // recieves appropriate input value.
```

Comment:
So, I'm not sure I like this approach very much. I understand it makes things cleaner in that all ops can retrieve their inputs naturally as they currently do via `nodeValueByName_`. However as you have had to implement it requires some sort of fixup via the `nonConstantsNodeValueMap`. I wonder if instead we could e.g. modify `getNodeValueByName()`, so that it checks if it should insert a ConvertToNode to FP16 before returning the NodeValue. Then we never modify `nodeValueByName_` and don't require the fixup. Additionally, CSE during graph optimizations should combine many ConvertToNodes that all use the same op. WDYT?

Reply:
Hi @jfix71,
I agree with you that changing the getNodeValueByName() and getConstantByName() we can avoid the fixup needed for nodes with multiple users via nonConstantsNodeValueMap. To handle operator output precision I can think of two approaches

1.  Modify addNodeAsOutput() method to add ConvertToFP32 node to its outputs and update nodeValueByName_.
2.  In getNodeValueByName() check if operator output name is not specified to run in fp16 and input is coming from ConvertToFP16 node, then add ConvertToFP32 and return the NodeValue. So getNodeValueByName() will return three possible NodeValues

              -    Original NodeValue
              -    Add ConvertToFP16 node and return the NodeValue if operator is needed to be in FP16 precision
              -    Add ConvertToFP32 node if the operator is needed to be in FP32 precision but the input is coming from a Node set to FP16 precision

Kinldy provide your suggestions.